pkgs/operator/state.go

```diff
@@ -220,11 +309,139 @@ func (s *Switch) InitInstance(reqID [24]byte, initMsg *wire.Transport, initiator
+	}
+	err = crypto.VerifyRSA(initiatorPubKey, marshalledWireMsg, initiatorSignature)
+	if err != nil {
+		return nil, fmt.Errorf("init message: initiator signature isn't valid: %s", err.Error())
+	}
+	initiatorID := sha256.Sum256(initiatorPubKey.N.Bytes())
+	s.Logger.Info("✅ reshare message signature is successfully verified", zap.String("from initiator", fmt.Sprintf("%x", initiatorID[:])))
+
+	s.Logger.Info("Starting resharing protocol")
+	// try to get old share local owner first
+	var shareFromDB basedb.Obj
+	secret := &kyber_dkg.DistKeyShare{}
+	shareFromDB, ok, err := s.DB.Get([]byte("secret"), reshare.OldID[:])
+	if err != nil {
+		return nil, err
+	}
+	if ok {
+		decBin, err := s.DecryptSecretDB(shareFromDB.Value)
+		if err != nil {
+			return nil, err
+		}
+		var privShare dkg.DistKeyShare
+		err = privShare.Decode(decBin)
+		if err != nil {
+			return nil, err
+		}
+		var coefs []kyber.Point
+		coefsBytes := utils.SplitBytes(privShare.Commits, 48)
+		for _, c := range coefsBytes {
+			p := bls3.NewBLS12381Suite().G1().Point()
+			err := p.UnmarshalBinary(c)
+			if err != nil {
+				return nil, err
+			}
+			coefs = append(coefs, p)
+		}
+		secret.Commits = coefs
+		secretPoint := bls3.NewBLS12381Suite().G1().Scalar()
+		err = secretPoint.UnmarshalBinary(privShare.Share.V)
+		if err != nil {
+			return nil, err
+		}
+		secret.Share = &share.PriShare{V: secretPoint, I: privShare.Share.I}
+		s.Mtx.Lock()
+		l := len(s.Instances)
+		if l >= MaxInstances {
+			cleaned := s.CleanInstances() // not thread safe
+			if l-cleaned >= MaxInstances {
+				s.Mtx.Unlock()
+				return nil, ErrMaxInstances
+			}
+		}
+		_, ok = s.Instances[reqID]
+		if ok {
+			tm := s.InstanceInitTime[reqID]
+			if !time.Now().After(tm.Add(MaxInstanceTime)) {
+				s.Mtx.Unlock()
+				return nil, ErrAlreadyExists
+			}
+			delete(s.Instances, reqID)
+			delete(s.InstanceInitTime, reqID)
+		}
+		s.Mtx.Unlock()
+		inst, resp, err := s.CreateInstanceReshare(reqID, reshare, initiatorPubKey, secret)
+		if err != nil {
+			return nil, err
+		}
+		s.Mtx.Lock()
+		_, ok = s.Instances[reqID]
+		if ok {
+			s.Mtx.Unlock()
+			return nil, ErrAlreadyExists
+		}
+		s.Instances[reqID] = inst
+		s.InstanceInitTime[reqID] = time.Now()
+		s.Mtx.Unlock()
+		return resp, nil
+	}
+	s.Mtx.Lock()
+	l := len(s.Instances)
+	if l >= MaxInstances {
+		cleaned := s.CleanInstances() // not thread safe
+		if l-cleaned >= MaxInstances {
+			s.Mtx.Unlock()
+			return nil, ErrMaxInstances
+		}
+	}
+	_, ok = s.Instances[reqID]
+	if ok {
+		tm := s.InstanceInitTime[reqID]
+		if !time.Now().After(tm.Add(MaxInstanceTime)) {
+			s.Mtx.Unlock()
+			return nil, ErrAlreadyExists
+		}
+		delete(s.Instances, reqID)
+		delete(s.InstanceInitTime, reqID)
+	}
+	s.Mtx.Unlock()
+	inst, resp, err := s.CreateInstanceReshare(reqID, reshare, initiatorPubKey, nil)
+	if err != nil {
+		return nil, err
+	}
+	s.Mtx.Lock()
+	_, ok = s.Instances[reqID]
+	if ok {
+		s.Mtx.Unlock()
+		return nil, ErrAlreadyExists
+	}
```

Comment:
there's a bug here. in this case the instance will be created and run but not returned. at least return the exiting one.

Reply:
this part is redundant, removed